### PR TITLE
[query] cleanup redundent MethodBuilders in Emit

### PR DESCRIPTION
**Pull request by patrick-schultz**
Emit was taking two MethodBuilder arguments, one directly and one embedded in an EmitRegion. This PR replaces the EmitRegion argument with a Value[Region]. It also makes a couple other simple refactorings that were personally helpful in understanding the structure of Emit.

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 1]
```diff
@@ -2500,18 +2508,19 @@ object NDArrayEmitter {
...
-abstract class NDArrayEmitter(
-   val nDims: Int,
-   val outputShape: IndexedSeq[Value[Long]],
-   val outputShapePType: PTuple,
-   val outputElementPType: PType,
-   val setupShape: Code[Unit],
-   val setupMissing: Code[Unit] = Code._empty,
-   val missing: Code[Boolean] = false) {
+abstract class NDArrayEmitter[C](
+  mb: EmitMethodBuilder[C],
```
Comment (johnc1231) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala:2512]: Why is this necessary? Could we not just make the `emit` function below take a `EmitMethodBuilder[C] argument instead of `EmitMethodBuilder[_]`?

Comment (johnc1231) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala:2512]: I'm struggling to find it now, but I'm pretty sure @cseed had a comment somewhere about `NDArrayEmitter` holding on to a `MethodBuilder` being a mistake

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala:2512]: yeah, emitLoops makes this wrong, I think

Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala:2512]: I don't remember exactly what made me do that. Probably it's not necessary, I was just trying to clarify things for myself. I can move it back. I'd like to know the argument against it.

Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala:2512]: emitLoops is only called once, with the `mb` used to construct the NDArrayEmitter. What is wrong with removing the EmitMethodBuilder argument from emitLoops?

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 2]
```diff
@@ -534,35 +535,39 @@ private class Emit[C](
...
     def wrapToMethod(irs: Seq[IR], mb: EmitMethodBuilder[C] = mb, env: E = env, container: Option[AggContainer] = container)(useValues: (EmitMethodBuilder[C], PType, EmitCode) => Code[Unit]): Code[Unit] =
       this.wrapToMethod(irs, mb, env, container)(useValues)
 
-    def emitStream(ir: IR, mb: EmitMethodBuilder[C] = mb): COption[EmitStream.SizedStream] =
-      EmitStream(this, mb, ir, env, er, container)
+    def emitStream(ir: IR): COption[EmitStream.SizedStream] =
+      EmitStream.emit(this, ir, mb, region, env, container)
 
-    def emitDeforestedNDArray(ir: IR, mb: EmitMethodBuilder[C] = mb, er: EmitRegion = er): EmitCode =
-      deforestNDArray(mb, er, ir, env).emit(mb, coerce[PNDArray](ir.pType))
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala:558]: I think this is probably what made me try moving the MethodBuilder argument to the NDArrayEmitter constructor. We pass `mb` both to `deforestNDArray`, which returns an `NDArrayEmitter`, and to `emit`. Are those MethodBuilders allowed to be different? I admit I don't understand the NDArrayEmitter very well.

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala (hunk 1 above)
Comment (cseed) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala:2512]: > I'm struggling to find it now, but I'm pretty sure @cseed had a comment somewhere about NDArrayEmitter holding on to a MethodBuilder being a design mistake

I don't really have a specific reason.  I thought they should be passed to methods that emit code in them, and not in data structures or anywhere else.  Just seemed less error prone.  But data structures that carry Code[_] might already be implicitly capturing one (e.g. through bound locals) so maybe it isn't such a big deal.

Mostly this is all moot when we stop splitting during emit, so I don't care too much.

I think things emitting code should have/take the method builder they're emitting into.  Eventually this will become a {Emit}CodeBuilder which will basically be the only way to emit code.

Comment (cseed) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala:2512]: I guess the motivation is that you might want to emit the loop in a different method than the one that's available when the NDArrayEmitter is constructed.  I thought that happens?

Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala:2512]: >  I thought they should be passed to methods that emit code in them, and not in data structures or anywhere else.

I agree with that. I just think of Emit and NDArrayEmitter as more like functions than data structures. They're roots of recursive call trees, and hold values in fields that are constant throughout the tree.

 But I also don't care too much. I'm just trying to understand (and reflect in the code) where MethodBuilders are required to be the same.

Comment (cseed) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala:2512]: My memory might be wrong/out of date, but I thought each emitLoop created a function to emit into, passing the created function down the chain, so the method builder(s) are not constant here.  I think when I ripped them out as John says, I did it because the one stored in the EmitNDArray was the wrong one for actually emitting the loops and I took it because it wasn't used otherwise and I didn't want it to be used accidentally.  But if the tests pass, I don't really care.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala]: > My memory might be wrong/out of date, but I thought each emitLoop created a function to emit into, passing the created function down the chain

Yes, that's how it works. We have some bugs related to which MB was used to emit body code inside of loops, though, which John is fixing now.

Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala]: Thanks for the explanation. I was missing where emitLoops passes its function down.
